leakcanary-haha/src/main/java/leakcanary/HprofReader.kt

```diff
@@ -0,0 +1,397 @@
+    val classId = readId()
+    val remainingBytesInInstance = readInt()
+    val fieldValues = readByteArray(remainingBytesInInstance)
+
+    return InstanceDumpRecord(
+        id = id,
+        stackTraceSerialNumber = stackTraceSerialNumber,
+        classId = classId,
+        fieldValues = fieldValues
+    )
+  }
+
+  fun readClassDumpRecord(
+    id: Long
+  ): ClassDumpRecord {
+    // stack trace serial number
+    val stackTraceSerialNumber = readInt()
+    val superClassId = readId()
+    // class loader object ID
+    val classLoaderId = readId()
+    // signers object ID
+    val signersId = readId()
+    // protection domain object ID
+    val protectionDomainId = readId()
+    // reserved
+    readId()
+    // reserved
+    readId()
+
+    // instance size (in bytes)
+    // Useful to compute retained size
+    val instanceSize = readInt()
+
+    // Skip over the constant pool
+    val constantPoolCount = readUnsignedShort()
+    for (i in 0 until constantPoolCount) {
+      // constant pool index
+      skip(SHORT_SIZE)
+      skip(typeSize(readUnsignedByte()))
+    }
+
+    val staticFields = mutableListOf<StaticFieldRecord>()
+    val staticFieldCount = readUnsignedShort()
+    for (i in 0 until staticFieldCount) {
+
+      val nameStringId = readId()
+      val type = readUnsignedByte()
+      val value = readValue(type)
+
+      staticFields.add(
+          StaticFieldRecord(
+              nameStringId = nameStringId,
+              type = type,
+              value = value
+          )
+      )
+    }
+
+    val fields = mutableListOf<FieldRecord>()
+    val fieldCount = readUnsignedShort()
+    for (i in 0 until fieldCount) {
+      fields.add(FieldRecord(nameStringId = readId(), type = readUnsignedByte()))
+    }
+
+    return ClassDumpRecord(
+        id = id,
+        stackTraceSerialNumber = stackTraceSerialNumber,
+        superClassId = superClassId,
+        classLoaderId = classLoaderId,
+        signersId = signersId,
+        protectionDomainId = protectionDomainId,
+        instanceSize = instanceSize,
+        staticFields = staticFields,
+        fields = fields
+    )
+  }
+
+  fun readObjectArrayDumpRecord(
+    id: Long
+  ): ObjectArrayDumpRecord {
+    // stack trace serial number
+    val stackTraceSerialNumber = readInt()
+    val arrayLength = readInt()
+    val arrayClassId = readId()
+    val elementIds = readIdArray(arrayLength)
+    return ObjectArrayDumpRecord(
+        id = id,
+        stackTraceSerialNumber = stackTraceSerialNumber,
+        arrayClassId = arrayClassId,
+        elementIds = elementIds
+    )
+  }
+
+  fun readPrimitiveArrayDumpRecord(
+    id: Long
+  ): PrimitiveArrayDumpRecord {
+    val stackTraceSerialNumber = readInt()
+    // length
+    val arrayLength = readInt()
+    val type = readUnsignedByte()
+    return when (type) {
+      BOOLEAN_TYPE -> BooleanArrayDump(
+          id, stackTraceSerialNumber, readBooleanArray(arrayLength)
```

Comment:
> This has the same advantage as mentioned above of sharing the backing byte[].

I didn't see another comment mentioning sharing the backing byte[]. Each ByteString has its own separate backing array so I'm not sure where the sharing occurs.

I do think it's an interesting optimization idea, but I ran some tests and because we're now skipping primitive arrays based on the index when traversing the graph, we actually never read those except when we do want to display them.

Added this:

```
      /**
       * Note: we could move the arrays to the parent class as a ByteString or ByteArray
       * and then each subtype can create a new array of the right type if needed.
       * However, experimenting with live parsing has shown that we never to read arrays except
       * when we want to display leak trace information, in which case we do need the data.
       */
      sealed class PrimitiveArrayDumpRecord : ObjectRecord() {
`